pallets/foreign-investments/src/impls/invest.rs

```diff
@@ -0,0 +1,1087 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge Chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use core::cmp::Ordering;
+
+use cfg_traits::SimpleCurrencyConversion;
+use cfg_types::investments::Swap;
+use sp_runtime::{
+	traits::{EnsureAdd, EnsureSub, Zero},
+	ArithmeticError, DispatchError,
+};
+
+use crate::types::{InvestState, InvestStateConfig, InvestTransition};
+
+impl<T> InvestState<T>
+where
+	T: InvestStateConfig,
+	/* impl<Balance, Currency> InvestState<Balance, Currency>
+	 * where
+	 * 	Balance: Clone + Copy + EnsureAdd + EnsureSub + Ord + Debug,
+	 * 	Currency: Clone + Copy + PartialEq + Debug, */
+{
+	/// Solely apply state machine to transition one `InvestState` into another
+	/// based on the transition, see <https://centrifuge.hackmd.io/IPtRlOrOSrOF9MHjEY48BA?view#State-diagram>.
+	///
+	/// NOTE: MUST call `apply_invest_state_transition` on the post state to
+	/// actually mutate storage.
+	pub fn transition(
+		&self,
+		transition: InvestTransition<T::Balance, T::CurrencyId>,
+	) -> Result<Self, DispatchError> {
+		match transition {
+			InvestTransition::IncreaseInvestOrder(swap) => Self::handle_increase(self, swap),
+			InvestTransition::DecreaseInvestOrder(swap) => Self::handle_decrease(self, swap),
+			InvestTransition::FulfillSwapOrder(swap) => {
+				Self::handle_fulfilled_swap_order(self, swap)
+			}
+			InvestTransition::CollectInvestment(amount_unprocessed) => {
+				Self::handle_collect(self, amount_unprocessed)
+			}
+		}
+	}
+
+	/// Returns the potentially existing active swap amount denominated in pool
+	/// currency:
+	/// * If the state includes `ActiveSwapIntoPoolCurrency`, it returns
+	///   `Some(swap.amount)`.
+	/// * If the state includes `ActiveSwapIntoForeignCurrency`, it returns
+	///   `Some(swap.amount)` converted into pool currency denomination.
+	/// * Else, it returns `None`.
+	pub(crate) fn get_active_swap_amount_pool_denominated(
+		&self,
+	) -> Result<T::Balance, DispatchError> {
+		match *self {
+			InvestState::NoState => Ok(T::Balance::zero()),
+			InvestState::InvestmentOngoing { .. } => Ok(T::Balance::zero()),
+			InvestState::ActiveSwapIntoPoolCurrency { swap } => Ok(swap.amount),
+			InvestState::ActiveSwapIntoForeignCurrency { swap } => Ok(T::CurrencyConverter::stable_to_stable(swap.currency_out, swap.currency_in, swap.amount)?),
```

Comment:
I think the order of the arguments provided to `stable_to_stable` is incorrect - I think it should be `swap.currency_in, swap.currency_out, swap.amount`. 

I noticed the same in other places where `stable_to_stable` is called.

Can you please double-check this? <3

Reply:
I think I might just be a bit confused by these calls...